sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableConfigTranslator.java

```diff
@@ -0,0 +1,342 @@
+import com.google.cloud.bigtable.data.v2.stub.BigtableBatchingCallSettings;
+import io.grpc.internal.GrpcUtil;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.security.KeyStore;
+import java.security.PrivateKey;
+import java.util.Objects;
+import javax.annotation.Nonnull;
+import org.apache.beam.sdk.extensions.gcp.options.GcpOptions;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.options.ValueProvider;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Strings;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableMap;
+import org.threeten.bp.Duration;
+
+/** Helper class to translate {@link BigtableOptions} to Bigtable Veneer settings. */
+@SuppressWarnings({
+  "nullness" // TODO(https://github.com/apache/beam/issues/20497)
+})
+class BigtableConfigTranslator {
+
+  /** Translate BigtableConfig and BigtableReadOptions to Veneer settings. */
+  static BigtableDataSettings translateReadToVeneerSettings(
+      @Nonnull BigtableConfig config,
+      @Nonnull BigtableReadOptions options,
+      @Nonnull PipelineOptions pipelineOptions) {
+    BigtableDataSettings.Builder settings = buildBigtableDataSettings(config, pipelineOptions);
+    return configureReadSettings(settings, options);
+  }
+
+  /** Translate BigtableConfig and BigtableWriteOptions to Veneer settings. */
+  static BigtableDataSettings translateWriteToVeneerSettings(
+      @Nonnull BigtableConfig config,
+      @Nonnull BigtableWriteOptions options,
+      @Nonnull PipelineOptions pipelineOptions) {
+
+    BigtableDataSettings.Builder settings = buildBigtableDataSettings(config, pipelineOptions);
+    return configureWriteSettings(settings, options);
+  }
+
+  private static BigtableDataSettings.Builder buildBigtableDataSettings(
+      BigtableConfig config, PipelineOptions pipelineOptions) {
+    BigtableDataSettings.Builder dataBuilder;
+    if (!Strings.isNullOrEmpty(config.getEmulatorHost())) {
+      String hostAndPort = config.getEmulatorHost();
+      try {
+        int lastIndexOfCol = hostAndPort.lastIndexOf(":");
+        int port = Integer.parseInt(hostAndPort.substring(lastIndexOfCol + 1));
+        dataBuilder =
+            BigtableDataSettings.newBuilderForEmulator(
+                hostAndPort.substring(0, lastIndexOfCol), port);
+      } catch (NumberFormatException | IndexOutOfBoundsException ex) {
+        throw new RuntimeException("Invalid host/port in BigtableConfig " + hostAndPort);
+      }
+    } else {
+      dataBuilder = BigtableDataSettings.newBuilder();
+    }
+
+    // Configure target
+    dataBuilder
+        .setProjectId(Objects.requireNonNull(config.getProjectId().get()))
+        .setInstanceId(Objects.requireNonNull(config.getInstanceId().get()));
+    if (config.getAppProfileId() != null) {
+      dataBuilder.setAppProfileId(Objects.requireNonNull(config.getAppProfileId().get()));
```

Comment:
What happens if the value in the ValueProvider is not set? I think your if statement needs to check the value as well: if (config.getAppProfileId() != null && !Strings.isNotNullOrEmpty(config.getAppProfileId()))